src/state/url-sync-mixin.ts

```diff
@@ -46,131 +36,32 @@ export const urlSyncMixin = <
-              historyPromise = new Promise((resolve) => {
-                historyResolve = () => {
-                  resolve();
-                  historyResolve = undefined;
-                  historyPromise = undefined;
-                };
-                mainWindow.history.back();
-              });
-            }
-          }
-        };
-
         private _popstateChangeListener = (ev: PopStateEvent) => {
-          if (this._ignoreNextPopState) {
-            if (
-              history.length &&
-              (ev.state?.oldState?.replaced ||
-                ev.state?.oldState?.dialogParams === null)
-            ) {
-              // if the previous dialog was replaced, or we could not copy the params, and the current dialog is closed, we should also remove the previous dialog from history
-              if (DEBUG) {
-                console.log("remove old state", ev.state.oldState);
-              }
-              mainWindow.history.back();
-              return;
-            }
-            if (DEBUG) {
-              console.log("ignore popstate");
-            }
-            this._ignoreNextPopState = false;
-            if (historyResolve) {
-              historyResolve();
-            }
-            return;
-          }
-          if (ev.state && "dialog" in ev.state) {
+          if (ev.state) {
             if (DEBUG) {
               console.log("popstate", ev);
             }
-            this._handleDialogStateChange(ev.state);
-          }
-          if (historyResolve) {
-            historyResolve();
-          }
-        };
-
-        private async _handleDialogStateChange(state: DialogState) {
-          if (DEBUG) {
-            console.log("handle state", state);
-          }
-          if (!state.open) {
-            const closed = await closeDialog(state.dialog);
-            if (!closed) {
-              if (DEBUG) {
-                console.log("dialog could not be closed");
-              }
-              // dialog could not be closed, push state again
-              mainWindow.history.pushState(
-                {
-                  dialog: state.dialog,
-                  open: true,
-                  dialogParams: null,
-                  oldState: null,
-                },
-                ""
-              );
-              return;
+            if (ev.state.opensDialog) {
+              // coming back from a dialog
+              // if we are instead navigating forward, the dialogs are already closed
+              closeLastDialog();
             }
-            if (state.oldState) {
-              if (DEBUG) {
-                console.log("handle old state");
-              }
-              this._handleDialogStateChange(state.oldState);
+            if ("dialog" in ev.state) {
+              // coming to a dialog
+              // in practice the dialog stack is empty when navigating forward, so this is a no-op
+              showDialogFromHistory(ev.state.dialog);
```

Comment:
It is not really a no-op, as it will trigger a `history.back` (causing the forward button to be enabled again) that will then trigger a `closeLastDialog` which will no-op

Reply:
Yeah, I meant to say nothing will happen from a user perspective but no-op was shorter :D